rts/motoko-rts/src/memory/ic.rs

```diff
@@ -38,6 +38,8 @@ unsafe extern "C" fn init(align: bool) {
         get_heap_base()
     };
     LAST_HP = HP;
+    // make sure that the page HP points to is allocated too
+    grow_memory(u64::from(HP))
```

Comment:
I wonder if we could avoid the need for this pre-allocation by using a separate global, `last_grow_memory_ptr`, as a cache of the last `ptr` passed to grow_memory. perhaps initialized to 0xFFFF_0000.
and then replacing the test to grow at line 80 with (something like)

```
if (last_grow_memory(ptr) ^ new_hp) / WASM_PAGE_SIZE.as_u32() as u64 != 0 {
            // Page boundary crossed, make sure that pages are allocated
            grow_memory(new_hp);
            last_grow_memory(ptr) = new_hp;
        }
```
Or would that allocate pages too late?

Reply:
But this line is almost never visited. I don't think we have to optimise this.

Reply:
It wasn't so much to optimize it but to avoid using the heap_ptr itself as the cach and to avoid preallocating the first page (if this works at all). No biggie though.

Reply:
It might be worth it. It needs to be initialised to 0x0000_FFFF and then a simple comparison against `new_hp` would suffice. Also the global would be monotonic and thus would avoid many calls to `grow_memory` when `HP` get decreased by GC. Thanks for the idea, I'll try this in a separate follow-up.

Reply:
Yes,I was actually wondering if it would be even better to just compare against the high-water mark (didn't realise this suggestion might also give you that.)

Reply:
Err, why don't we just compare against memorysize × 65536 directly? No need for a cache unless expensive.

Reply:
FYI #3790 does the experiment of not pre-allocating the first page and testing against memory size instead of the xor trick.